### DISCUSSION TITLE
Python dependency management

### DISCUSSION THREAD

**Discussion by dhruvkb** [mathesar-foundation/mathesar#8] [Maintainer discussions]
Should we use [Pipenv][pipenv] or [Poetry][poetry] to manage Python dependencies in the project? If yes, which?

[pipenv]: https://pipenv.pypa.io/en/latest/
[poetry]: https://python-poetry.org

**Comment by dhruvkb** [comment]
Personally, I lean towards yes and Pipenv (but that's just because I've never used Poetry so far).

**Reply by dhruvkb** [reply]
As a proof of concept, I set the project up with Pipenv in a fork [here](https://github.com/dhruvkb/mathesar/tree/pipenv).

**Reply by mathemancer** [reply]
I'm slightly biased against `pipenv` based on the fact that their emoji usage can make some terminals crash in some (rare) cases.  That said, I haven't used `poetry` at work before, just experimented with some personal projects, but I _do_ like that it works properly with the `pyproject.toml` (though there may be ways to merge the `Pipfile` with that easily if desired).

**Reply by dhruvkb** [reply]
If you're main complaint against Pipenv is the use of emojis, [that can be turned off](https://pipenv.pypa.io/en/latest/advanced/#pipenv.environments.PIPENV_HIDE_EMOJIS) with an [env var](https://github.com/dhruvkb/mathesar/commit/017f183).

**Comment by kgodey** [comment]
Between the two, I'd lean towards Poetry, despite also never having used it before. See https://github.com/python-poetry/poetry#what-about-pipenv. I've had some issues with pipenv being really slow, so have others, see https://github.com/pypa/pipenv/issues/2873 and related issues.

See also https://chriswarrick.com/blog/2018/07/17/pipenv-promises-a-lot-delivers-very-little/.

However, I'm not convinced that we need to use either, we have `requirements.txt` and `pip`. Why use another tool at all?

**Reply by dhruvkb** [reply]
At the risk of sounding like an advocate of Pipenv, I use Pipenv regularly in all my Python projects, including Django + DRF based ones, and have not experienced any of the problems, so the cases they occur in would most likely be disjoint from mine.

The main reason I recommend a package-manager is mostly so that
- we need only list the top-level dependencies in the `Pipfile`
- the complete details of all frozen packages can be stored separately in `Pipfile.lock`

Comparing this to `requirements.txt`, we'll need to either manually write in the requirements file to ensure that only top-level dependencies are stored there (and risk package version differences in environments) or use `pip freeze` which will write all packages with exact versions (making it harder to upgrade packages).

**Comment by mathemancer** [comment]
A couple of points in favor of `pipenv`/`poetry` on top of just `pip`:
- hashes of dependencies for security (i.e., the downloads can be verified to be the same for a given version)
- removing packages along with the things they pulled in as dependencies is easier than with just `pip`

**Reply by mathemancer** [reply]
point two isn't as relevant in our use case, though, since we'll have a fresh build in the docker container (or a fresh virtualenv)

**Reply by dhruvkb** [reply]
I would not say discard point two as irrelevant to our case. If we remove a package, we might still be leaving it's dependencies behind in the requirements generated by `pip freeze` (unless we write the `requirements.txt` file by hand).

**Reply by mathemancer** [reply]
Valid point

**Reply by kgodey** [reply]
I did a bit more research and we should use `pip-tools`: https://github.com/jazzband/pip-tools. We'll need to maintain a `requirements.in` by hand, but I think that's a good thing since it's best to be very intentional about dependencies. 

Here's a good article comparing the options: https://hynek.me/articles/python-app-deps-2018/

**Comment by silentninja** [comment]
I am leaning towards poetry since I have been using poetry and haven't faced any major problems so far. I like pip-tools too, would be a nice alternative if we don't need any advanced dependency management(doubtful, especially with plugins feature for mathesar)

 Mathesar has moving parts like csv module, db module(would prefer for it to be a standalone library) suitable to be separate from the core and with more modular features coming in, a mono repo is much suitable for the project. Compared to other dependency managers, monorepo support is actively [considered](https://github.com/python-poetry/poetry/issues/936) and supported by poetry.
